### PR TITLE
Add support for has_many_aggregate when working with STI models

### PR DESCRIPTION
**Pull request by gfieldclio**
When using STI, `has_many_aggregate` to aggregate child models, under certain conditions, will fail to find the existing results. This is because the query uses an alias for the table join (since it's joining to the same table), but the `jit_preloader` logic naively uses the table's name.

These changes add logic to determine when a table alias is present in the query, and when one is, use that for the `jit_preloader` query.

While looking into this issue, we also discovered that the aggregate query fails to limit the results based on type when querying a child model through another child model. These changes also address that issue by including the type condition when querying through a child model.

### Before:
```sql
SELECT COUNT(*) AS count_all, contacts.contact_book_id AS contacts_contact_book_id
FROM "contacts"
INNER JOIN "contacts" "companies_company_employees" ON "contacts"."contact_id" = "companies_company_employees"."id"
WHERE "contacts"."type" IN ('Employee') AND "contacts"."contact_book_id" = ?
GROUP BY contacts.contact_book_id
```

### After:
```sql
SELECT COUNT(*) AS count_all, companies_company_employees.contact_book_id AS companies_company_employees_contact_book_id
FROM "contacts"
INNER JOIN "contacts" "companies_company_employees" ON "contacts"."contact_id" = "companies_company_employees"."id"
WHERE "contacts"."type" IN ('Employee') AND "companies_company_employees"."contact_book_id" = ? AND "companies_company_employees"."type" = ?
GROUP BY companies_company_employees.contact_book_id
```

### CODE REVIEW

File: lib/jit_preloader/active_record/base.rb [hunk 1]
```diff
@@ -41,11 +41,22 @@ def has_many_aggregate(assoc, name, aggregate, field, default: 0)
             association_scope = klass.all.merge(association(assoc).scope).unscope(where: aggregate_association.foreign_key)
             association_scope = association_scope.instance_exec(&reflection.scope).reorder(nil) if reflection.scope
 
-            conditions[aggregate_association.table_name] = { aggregate_association.foreign_key => primary_ids }
+            # If the query uses an alias for the association, use that instead of the table name
+            table_alias_name = association_scope.references_values.first || aggregate_association.table_name
```
Comment (doliveirakn) [inline lib/jit_preloader/active_record/base.rb]: I'm not convinced that `association_scope.references_values.first` be the value you are looking for. 

If the `association_scope` happens to have multiple references this may be incorrect.

```
association_scope.references(:contact).references_values
# => ["contacts_employees", "contact"]
```

How can we ensure that we are always referencing the correct table alias (in particular if there are multiple references)

File: lib/jit_preloader/active_record/base.rb [hunk 2]
```diff
@@ -41,11 +41,22 @@ def has_many_aggregate(assoc, name, aggregate, field, default: 0)
...
+            conditions[table_alias_name] = { aggregate_association.foreign_key => primary_ids }
+
+            # If the association is a STI child model, specify its type in the condition so that it
+            # doesn't include results from other child models
+            is_base_class = aggregate_association.klass.superclass.abstract_class? || aggregate_association.klass.superclass == ActiveRecord::Base
+            has_type_column = aggregate_association.klass.column_names.include?(aggregate_association.klass.inheritance_column)
+            if is_child_sti_model = !is_base_class && has_type_column
```
Comment (doliveirakn) [inline lib/jit_preloader/active_record/base.rb]: The `is_child_sti_model` is not used anywhere else and makes this boolean logic hard to follow

File: lib/jit_preloader/active_record/base.rb [hunk 3]
```diff
@@ -41,11 +41,22 @@ def has_many_aggregate(assoc, name, aggregate, field, default: 0)
...
+              conditions[table_alias_name].merge!({ type: aggregate_association.klass.sti_name })
```
Comment (doliveirakn) [inline lib/jit_preloader/active_record/base.rb]: We shouldn't hard code `type` here. 

`type:` => `aggregate_association.klass.inheritance_column`.

File: lib/jit_preloader/active_record/base.rb [hunk 4]
```diff
@@ -41,11 +41,22 @@ def has_many_aggregate(assoc, name, aggregate, field, default: 0)
...
 
-            conditions[aggregate_association.table_name] = { aggregate_association.foreign_key => primary_ids }
+            # If the query uses an alias for the association, use that instead of the table name
+            table_alias_name = association_scope.references_values.first || aggregate_association.table_name
+            conditions[table_alias_name] = { aggregate_association.foreign_key => primary_ids }
+
+            # If the association is a STI child model, specify its type in the condition so that it
+            # doesn't include results from other child models
+            is_base_class = aggregate_association.klass.superclass.abstract_class? || aggregate_association.klass.superclass == ActiveRecord::Base
+            has_type_column = aggregate_association.klass.column_names.include?(aggregate_association.klass.inheritance_column)
```
Comment (doliveirakn) [inline lib/jit_preloader/active_record/base.rb]: I thinks all of this could be replaced with
```
if aggregate_association.klass != aggregate_association.klass.base_class
  # At this point I believe you know that `aggregate_association.klass` is an STI table and itsn't the base class
end
```

File: spec/lib/jit_preloader/preloader_spec.rb [hunk 5]
```diff
@@ -49,6 +48,51 @@
     ->(event, data){ source_map[data[:source]] << data[:association] }
   end
 
+  context "for single table inheritance" do
+    let!(:contact_book) { ContactBook.create(name: "The Yellow Pages") }
```
Comment (doliveirakn) [inline spec/lib/jit_preloader/preloader_spec.rb]: This is an awful lot of data for a spec. Do we actually need all of it?

If so, let's structure the tests so that tests only use the pieces of data that they need. It doesn't look like every spec needs all of the data.

File: lib/jit_preloader/active_record/base.rb (hunk 1 above)
Comment (doliveirakn) [reply lib/jit_preloader/active_record/base.rb]: Not perfect either, but we can also get some information form the `joins_values`

```
association_scope.joins_values.first.left.right
# => "companies_company_employees"
```

Though this doesn't entirely work with the specs. Just post here for reference while I was digging.

File: spec/lib/jit_preloader/preloader_spec.rb (hunk 5 above)
Comment (gfieldclio) [reply spec/lib/jit_preloader/preloader_spec.rb]: The data is used entirely, but each sub-context only uses part of it. I can move them within each context if you would prefer. I had avoided doing that as I didn't want to deviate too much from the existing spec structure.

File: lib/jit_preloader/active_record/base.rb (hunk 4 above)
Comment (gfieldclio) [reply lib/jit_preloader/active_record/base.rb]: Not necessarily. You can have one ActiveRecord model inherit from another without using STI. I'm not sure what the situation is where you would want to do this, but it's probably better to be safe here.

I was able to duplicate this by doing the following:
- Make a `PostalAddress` model that inherited from the `Address` model
- Update all of the `addresses` references on the other models to `postal_addresses`
- Update the `Contact` instances generated in the specs to use the `PostalAddress` model
- Add a debugger before the `if` statement
- Run this test https://github.com/clio/jit_preloader/blob/fe7cc4c543f29d04d9c61ca53556846ba7c3e029/spec/lib/jit_preloader/preloader_spec.rb#L276

File: lib/jit_preloader/active_record/base.rb (hunk 1 above)
Comment (gfieldclio) [reply lib/jit_preloader/active_record/base.rb]: I tried to simulate a case where there would be multiple aliases for a single table. For reference, I did the following:
- Added references 2 more levels deep for the employee relations to the `ContactBook` model
```
  has_many :company_employees_of_employees, through: :company_employees, source: :employees
  has_many :peons, through: :company_employees_of_employees, source: :employees
```
- Added an aggregation method for peons to the `ContactBook` model
```
  has_many_aggregate :peons, :count, :count, "*"
```
- Added data to the nested child through another child context (https://github.com/clio/jit_preloader/blob/93ccddd3fd42c81617b6ccbc671e63504cf589e9/spec/lib/jit_preloader/preloader_spec.rb#L79)
```
      let!(:company_employee_of_employee1) { Employee.create(name: "Company Employee of Employee1", contact: company_employee1) }
      let!(:company_employee_of_employee2) { Employee.create(name: "Company Employee of Employee2", contact: company_employee2) }
      let!(:peon1) { Employee.create(name: "Company Employee of Employee1", contact: company_employee_of_employee1) }
      let!(:peon2) { Employee.create(name: "Company Employee of Employee2", contact: company_employee_of_employee2) }
```
- Modified the spec in that context to use `peons_count` instead of `company_employees_count`
- Added a debugger before https://github.com/clio/jit_preloader/blob/fe07e37e6ae60ba34178be2796123739699f5622/lib/jit_preloader/active_record/base.rb#L46 and ran the spec

The result gave me the following:
```
(byebug) association_scope.references_values
["companies_peons"]
```

The query that it executed was:
```sql
SELECT COUNT(*) AS count_all, companies_peons.contact_book_id AS companies_peons_contact_book_id
FROM "contacts"
INNER JOIN "contacts" "company_employees_of_employees_peons" ON "contacts"."contact_id" = "company_employees_of_employees_peons"."id"
INNER JOIN "contacts" "company_employees_peons" ON "company_employees_of_employees_peons"."contact_id" = "company_employees_peons"."id"
INNER JOIN "contacts" "companies_peons" ON "company_employees_peons"."contact_id" = "companies_peons"."id"
WHERE "contacts"."type" IN ('Employee') AND "companies_peons"."contact_book_id" = ? AND "companies_peons"."type" = ?
GROUP BY companies_peons.contact_book_id
```

I'm not sure if this is conclusive enough to demonstrate that the current method works. I'll continue to dig, but I'm not sure what type of query would result in multiple results for `references_values`.

Comment (kevinhughes27) [reply lib/jit_preloader/active_record/base.rb]: @mctaylorpants and I stepped through this as well. Unfortunately Rails doesn't seem to provide a better way to detect this exact scenario but we also confirmed the results make sense and our test suite covers what we support.

We'd like to suggest a slight code change to better communicate the intention:

```ruby
table_alias_name = association_scope.references_values.first 
name = table_alias_name || aggregate_association.table_name

conditions[name] = { aggregate_association.foreign_key => primary_ids }
```

File: lib/jit_preloader/active_record/base.rb (hunk 4 above)
Comment (kevinhughes27) [reply lib/jit_preloader/active_record/base.rb]: Does this only cover the case where our class hierarchy is:
```
ActiveRecord::Base
  GenericParent
     ChildSTIModel
     ChildSTIModel2
```

What if there was another parent in between? Is this something we need to consider?

Comment (gfieldclio) [reply lib/jit_preloader/active_record/base.rb]: I believe the logic we have should cover cases where there is a parent in between. The logic is checking two things:
- That the direct parent of the class isn't an abstract class or `ActiveRecord::Base`
- That the class has the inheritance column defined on its DB table (which is how it identifies it as an STI model)

This logic should correctly identify an STI model in the case that you defined. That said, I think any child of an ActiveRecord model that has defined the inheritance column is an STI model.
